c2rust-analyze/src/labeled_ty.rs

```diff
@@ -220,6 +220,47 @@ impl<'tcx, L: Copy> LabeledTyCtxt<'tcx, L> {
         self.mk_slice(&tys.iter().map(|&ty| self.label(ty, f)).collect::<Vec<_>>())
     }
 
+    /// Substitute in arguments for any type parameter references (`Param`) in a labeled type.
+    /// Panics if `lty` contains a reference to a type parameter that is past the end of `substs`
+    /// (usually this means the caller is providing the wrong list of type arguments as `substs`).
+    ///
+    /// This produces a [`LabeledTy`] with the right structure and also substitutes the
+    /// underlying [`Ty`]s
+    pub fn subst_full(
+        &self,
+        lty1: LabeledTy<'tcx, L>,
+        lty2: LabeledTy<'tcx, L>,
+        substs: &[LabeledTy<'tcx, L>],
+    ) -> LabeledTy<'tcx, L> {
```

Comment:
This approach seems quite tricky - we should fix `subst` instead so that there's no need to mess with two different `LTy`s when doing a substitution.  I think the strategy for that is:

1. Convert `substs` to a `SubstsRef` (discarding labels/args and keeping only each `x.ty`)
2. Do the actual `Ty` substitution with `EarlyBinder(lty.ty).subst(...)` (based on [`subst_and_normalize_erasing_regions`](https://github.com/rust-lang/rust/blob/7c306f6dcd600b1fcc74439c780e8ebef338d84c/compiler/rustc_middle/src/ty/normalize_erasing_regions.rs#L133-L156))
3. Traverse and update `args` as is currently done in `LabeledTyCtxt::subst` (preferably with some `debug_assert`s in there to ensure the substituted `ty` has the expected shape to match the `args`)

Reply:
`subst` doesn't seem to be available, maybe I'm missing something:

```
no method named `subst` found for struct `EarlyBinder` in the current scope
   --> c2rust-analyze/src/labeled_ty.rs:238:37
    |
238 |                 *t = EarlyBinder(t).subst(self.tcx, substs);
    |                                     ^^^^^ method not found in `EarlyBinder<&mut rustc_middle::ty::Ty<'tcx>>`
```

I found its definition in `subst.rs`, though.

Reply:
There's a `Subst` trait you have to import.

Reply:
@spernsteiner let me know if 5db7e8d8 doesn't fit the requirement

Reply:
No longer used in d8876a7a